### PR TITLE
`forge script` Avoid bulk gasEstimation when setting `--skip-simulation`

### PR DESCRIPTION
**Pull request by ckoopmann**
## Motivation

Even when setting `--skip-simulation` the current implementation still tries to estimate gas for all transactions in a script before submitting any of them. 

Therefore scripts that have transactions that depend on previous tx's to succeed will fail. ( Avoiding which was one of the motivations to introduce `--skip-simulation` in the first place).

## Solution

When providing `--skip-simulation` don't try to estimate gas until immediately before tx submission. (If you combine this with the `--slow` option the state against which gas is estimated should include previous tx's from the same script, thus avoiding above outlined problem).

### CODE REVIEW

**Comment by ckoopmann** [comment]
Ready for review.
@onbjerg I commited the changed testcase into a separate commit (where it is failing) to illustrate the issue this pr is solving. Lmk if I should squash before merge. (or feel free to do so yourself).

**Review by mattsse — APPROVED** [review]
lgtm, defer to @joshieDo for final approval

File: cli/src/cmd/forge/script/broadcast.rs
Comment (onbjerg) [inline cli/src/cmd/forge/script/broadcast.rs:348]: is this caused by user error or by us? if by user error would prefer we return an `Error` instead of panic

File: cli/src/cmd/forge/script/broadcast.rs
Comment (onbjerg) [inline cli/src/cmd/forge/script/broadcast.rs:341]: Can you add context to this error? `.wrap_err_with(|| format("Failed to estimate gas bla bla bla"))?`

**Review by onbjerg — APPROVED** [review]
small stuff otherwise lgtm, defer to @joshieDo as well

File: cli/src/cmd/forge/script/broadcast.rs
Comment (ckoopmann) [reply cli/src/cmd/forge/script/broadcast.rs:348]: I just moved this part of the code around, so I'm not exactly sure in what scenario this panic would be thrown. However since this is very late in the "script lifecycle" (after off-chain / on-chain simulations and just before broadcasting the tx), I'd assume that a user error would have been caught earlier.

File: cli/src/cmd/forge/script/broadcast.rs
Comment (ckoopmann) [reply cli/src/cmd/forge/script/broadcast.rs:341]: Good point. I added the context inside the `estimate_gas` method to have it consistent across calls:
https://github.com/foundry-rs/foundry/pull/2601/files#diff-2b2997f85d225c8c6fdd337f5382a5be0fbd7f254f4d7854161576400646e563R375

**Review by mattsse — APPROVED** [review]
I think this is ready, @onbjerg?

**Comment by joshieDo** [comment]
lgtm, given the conflict fix

sorry was sick and my github notifications are a mess.

**Review by joshieDo — APPROVED** [review]

**Review by gakonst — APPROVED** [review]
LGTM - needs merge conflict to be resolved

**Comment by mattsse** [comment]
rebased, merging, after checks

**Pull request merged**
